### PR TITLE
Content blocks

### PR DESCRIPTION
**Pull request by itays123**
This is the content blocks backend implementation, as stated in issue #11

### CODE REVIEW

File: server/routes/blocks.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  const {
+    getBlockList,
+    getBlockById,
+    createBlock,
+    updateBlock,
+    deleteBlock,
+  } = require("../controllers/blocks");
+
+  app
+    .get("/api/blocks", populateUser, getBlockList)
```
Comment (davidmeirlevy) [inline server/routes/blocks.js:14]: Guests and regular users doesn’t need to have access to get list of all blocks.

File: server/models/block.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  const q = () =>
+    this.find(query)
+      .limit(limit)
+      .offset(offset)
+      .then((list) => {
+        if (list && list.length) return JSON.stringify(list);
+        else return "[]";
+      });
+
+  if (useCache) {
```
Comment (davidmeirlevy) [inline server/models/block.js:63]: Search blocks is a method that should be available to editors only from the admin panel.
No need to cache it.

File: server/controllers/blocks.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
+const Block = require("../models/block");
+
+function getBlockList(req, res) {
```
Comment (davidmeirlevy) [inline server/controllers/blocks.js:3]: It should be getBlocksList.. “Blocks”.

File: server/controllers/blocks.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  const reqQuery = { ...req.query || {} };
+
+  const isFrontTargeted = reqQuery.target === "front" ||
```
Comment (davidmeirlevy) [inline server/controllers/blocks.js:6]: This function should not be available for front use.. only from admin.

File: server/controllers/blocks.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  );
+  newBlock.save()
+    .then((block) => {
+      if (!block) return Promise.reject(null);
+      block = block.toObject();
+      res.status(200).json(block).end();
+    })
+    .catch((err) =>
+      res.status(400).json(
+        { message: err || "error while creating block, try again" },
```
Comment (davidmeirlevy) [inline server/controllers/blocks.js:58]: Avoid to pass the actual error to the frontend. It’s a security risk.

File: server/controllers/blocks.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  const { block: currBlock, body } = req;
+  delete body.tenant;
+
+  Object.assign(currBlock, body).save()
+    .then((newBlock) => {
+      res.status(200).json(newBlock).end();
+    })
+    .catch((err) =>
+      res.status(400).json(
+        { message: err || "error while updating block, try again" },
```
Comment (davidmeirlevy) [inline server/controllers/blocks.js:73]: Avoid to pass the actual error to the frontend. It’s a security risk.

File: server/controllers/blocks.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+function deleteBlock(req, res) {
+  const { block: deletedBlock } = req;
+
+  deletedBlock.remove()
+    .then((block) => {
+      res.status(200).json(block).end();
+    })
+    .catch((err) =>
+      res.status(400).json(
+        { message: err || "error while deleting block, try again" },
```
Comment (davidmeirlevy) [inline server/controllers/blocks.js:87]: Avoid to pass the actual error to the frontend. It’s a security risk.

File: server/models/block.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      "images", // for a gallery of images
+      "script", // for ads
+    ],
+  },
+});
+
+// some useful methods
+
+BlockSchema.statics.noScripts = function noScripts(useCache = false) {
+  const q = () =>
```
Comment (davidmeirlevy) [inline server/models/block.js:33]: If no one use it, remove it.

**Review by davidmeirlevy — CHANGES_REQUESTED** [review]

File: server/controllers/blocks.js (hunk 4 above)
Comment (itays123) [reply server/controllers/blocks.js:6]: onlyEditor or onlyAdmin?

Comment (davidmeirlevy) [reply server/controllers/blocks.js:6]: editor, from admin-panel.

Comment (itays123) [reply server/controllers/blocks.js:6]: Got it.

**Review by davidmeirlevy — APPROVED** [review]

**Pull request merged**